airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3StorageOperations.java

```diff
@@ -133,20 +136,30 @@ private String loadDataIntoBucket(final String objectPath, final SerializableBuf
     final long partSize = s3Config.getFormatConfig() != null ? s3Config.getFormatConfig().getPartSize() : DEFAULT_PART_SIZE;
     final String bucket = s3Config.getBucketName();
     final String fullObjectKey = objectPath + getPartId(objectPath) + getExtension(recordsData.getFilename());
+
+    final Map<String, String> metadata = new HashMap<>();
+    for (final BlobDecorator blobDecorator : blobDecorators) {
+      blobDecorator.updateMetadata(metadata, getMetadataMapping());
+    }
     final StreamTransferManager uploadManager = StreamTransferManagerHelper
-        .getDefault(bucket, fullObjectKey, s3Client, partSize)
+        .getDefault(bucket, fullObjectKey, s3Client, partSize, metadata)
         .checkIntegrity(true)
         .numUploadThreads(DEFAULT_UPLOAD_THREADS)
         .queueCapacity(DEFAULT_QUEUE_CAPACITY);
     boolean succeeded = false;
-    try (final MultiPartOutputStream outputStream = uploadManager.getMultiPartOutputStreams().get(0);
-        final InputStream dataStream = recordsData.getInputStream()) {
+
+    OutputStream outputStream = uploadManager.getMultiPartOutputStreams().get(0);
+    for (final BlobDecorator blobDecorator : blobDecorators) {
+      outputStream = blobDecorator.wrap(outputStream);
+    }
+    try (final InputStream dataStream = recordsData.getInputStream()) {
       dataStream.transferTo(outputStream);
       succeeded = true;
     } catch (final Exception e) {
       LOGGER.error("Failed to load data into storage {}", objectPath, e);
       throw new RuntimeException(e);
     } finally {
+      outputStream.close();
```

Comment:
if close fails, it won't execute the uploadManager cleanups

Reply:
fixed